### ISSUE TITLE
Add k-d tree index

### ISSUE THREAD

**Issue by benbovy** [ESM-VFC/xoak#8]
It would be interesting to test/benchmark the ball tree + haversine approach versus the kd-tree + conversion lat/lon spherical -> xyz cartersian.

The second approach is also used by libpysal, see [here](https://github.com/pysal/libpysal/blob/master/libpysal/cg/kdtree.py).

Since we already rely on scikit-learn for the ball tree, would it make sense to reuse scikit-learn's KDTree instead of Scipy's cKDTree?

**Comment by willirath** [comment]
> would it make sense to reuse scikit-learn's KDTree instead of Scipy's cKDTree?

Yes, definitely!

> It would be interesting to test/benchmark the ball tree + haversine approach versus the kd-tree + conversion lat/lon spherical -> xyz cartersian.

I'm not sure I understand why we should not also do the ball tree benchmark with lat/lon to spherical and the kd-tree with haversine. I'd go for a matrix like
```
```

**Comment by willirath** [comment]
> would it make sense to reuse scikit-learn's KDTree instead of Scipy's cKDTree?

Yes, definitely!

> It would be interesting to test/benchmark the ball tree + haversine approach versus the kd-tree + conversion lat/lon spherical -> xyz cartersian.

I'm not sure I understand why we should not also do the ball tree benchmark with lat/lon to spherical and the kd-tree with haversine. I'd go for a matrix like
```
tree:
  - ball tree
  - kd-tree
transform_metric:
  - (deg2rad, haversine)
  - (to sphere, euclidian)
```

(Sorry, hit send to early.)

**Comment by benbovy** [comment]
I don't think we can use k-d tree with lat/lon haversine: https://news.ycombinator.com/item?id=9281998

The rest of the matrix is valid. Ball tree works with any metric that respects the triangle inequality.

**Comment by benbovy** [comment]
Actually, there is already detailed comparison + benchmarks in [Jake's blog post](https://jakevdp.github.io/blog/2013/04/29/benchmarking-nearest-neighbor-searches-in-python/). But it'd still be interesting to see it applied to NEMO/FESOM spatial data patterns and the metrics/conversions used here.

**Comment by benbovy** [comment]
Mmm now I'm confused about the validity of kd-tree + haversine. I'm wondering why haversine is not in the list of valid metrics for scikit-learn's KDTree.

We should also consider the implementation in [pykdtree](https://github.com/storpipfugl/pykdtree).
